text/0523-model-argument-for-route-templates.md

```diff
@@ -0,0 +1,359 @@
+the least, but also serves no purpose at all.
+
+We opted to avoid going down that rabit hole and just treated it as a
+boilerplate syntax you have to type.
+
+If we are able to explain this using `@model` instead, it would make things a
+lot smoother here.
+
+We wouldn't have to go into the details of _how_ things get invoked internally.
+At the end of the day, it is still just a piece of syntax you type to get the
+job done, but at least it is consistent with the general mental model that
+`@names` means that thing was passed in, and it does not trigger the questions
+about what `this` is.
+
+It is also _not_ inconsistent with the fact that there is, in fact, a `this`
+context on the template. The usage and preference of using `@names` does not
+preclude the existance of a class (which provides the `this` context), as in
+the case of `@ember/components`.
+
+As for the concrete changes that need to be made in the documentation:
+
+* We will explain that `@model` is passed into the component from the route.
+  For example, in the `Route` class' `model` hook documentation, we can say
+  something like:
+
+  > The model hook is responsible for fetching and preparing any data that you
+  > need for your route. Ember will automatically call this hook for when
+  > entering a route, so that you can have an opportunity to do what you need to
+  > get the data you need. The object returned from this hook is known as the
+  > _model_ for the route.
+  >
+  > Note that since this is an `async` hook, if a promise is returned, it will
+  > be automatically `await`ed by Ember, and the _resolved value_ of the
+  > promise, as opposed to the promise object itself, will become the model of
+  > the route.
+  >
+  > The model object can be accessed from the route's template as using the
+  > `@model` argument. By default, the controller can also accessed the route's
+  > model from `this.model`. The latter behavior is customizable, see the
+  > _setupController_ method for details.
+  >
+  > The model of a route should be treated as read-only. For example, the
+  > controller _should not_ mutate its `this.model` property, as doing so will
+  > cause it to get out of sync with the rest of the system. Specifically,
+  > doing so will _not_ update the current URL, the `@model` argument in the
+  > template, the router service, etc. Instead, a _route transition_ should be
+  > performed.
+
+* Guides and API docs should be updated to prefer `@model` in route templates.
+* API docs will still document the `model` property.
+* We will remove any examples that uses `{{this.model}}` in templates, or replace them with `{{@model}}`. i.e. we won’t be documenting using `this.model` in templates, even thought it would “work”.
+
+## Drawbacks
+
+In some applications, developers have developed a pattern to override the
+`setupController` method to assign the model to a different property on the
+controller other than the default `model` naming convention.
```

Comment:
@chancancode by my reading of the detailed design, "`@model` will not be updated when `this.model` is mutated" which implied it is somehow coupled to the return from `model() {` and not based on the controller's `model` property.

As such why would the implementation of `setupController` matter? Shouldn't the return value of `model() {` *always* be available at `@model` regardless of what happens in `setupController`?

Reply:
Yep. That is correct. This is referring to that some teams currently prefer to override the hook to set `this.post` instead of `this.model`, because they prefer the name. They will still be able to do that, and `@model` will still contain the right value, they would just have to choose between going back to the model name (`@model`) or continue to indirect via the controller (`this.post`), not `@post` (yet).